src/query/api/v1/handler/placement/replace.go

```diff
@@ -0,0 +1,154 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package placement
+
+import (
+	"net/http"
+	"path"
+	"time"
+
+	"github.com/m3db/m3/src/cluster/placement"
+	"github.com/m3db/m3/src/query/api/v1/handler"
+	"github.com/m3db/m3/src/query/generated/proto/admin"
+	"github.com/m3db/m3/src/query/util/logging"
+	xhttp "github.com/m3db/m3/src/x/net/http"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"go.uber.org/zap"
+)
+
+const (
+	// ReplaceHTTPMethod is the HTTP method for the the replace endpoint.
+	ReplaceHTTPMethod = http.MethodPost
+
+	replacePathName = "replace"
+)
+
+var (
+	// M3DBReplaceURL is the url for the m3db replace handler (method POST).
+	M3DBReplaceURL = path.Join(handler.RoutePrefixV1, M3DBServicePlacementPathName, replacePathName)
+)
+
+// ReplaceHandler is the type for placement replaces.
+type ReplaceHandler Handler
+
+// NewReplaceHandler returns a new ReplaceHandler.
+func NewReplaceHandler(opts HandlerOptions) *ReplaceHandler {
+	return &ReplaceHandler{HandlerOptions: opts, nowFn: time.Now}
+}
+
+func (h *ReplaceHandler) ServeHTTP(serviceName string, w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	logger := logging.WithContext(ctx)
+
+	req, pErr := h.parseRequest(r)
+	if pErr != nil {
+		xhttp.Error(w, pErr.Inner(), pErr.Code())
+		return
+	}
+
+	placement, err := h.Replace(serviceName, r, req)
+	if err != nil {
+		status := http.StatusInternalServerError
+		if _, ok := err.(unsafeAddError); ok {
+			status = http.StatusBadRequest
+		}
+		logger.Error("unable to replace instance", zap.Any("error", err))
+		xhttp.Error(w, err, status)
+		return
+	}
+
+	placementProto, err := placement.Proto()
+	if err != nil {
+		logger.Error("unable to get placement protobuf", zap.Any("error", err))
+		xhttp.Error(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	resp := &admin.PlacementGetResponse{
+		Placement: placementProto,
+		Version:   int32(placement.GetVersion()),
+	}
+
+	xhttp.WriteProtoMsgJSONResponse(w, resp, logger)
+}
+
+func (h *ReplaceHandler) parseRequest(r *http.Request) (*admin.PlacementReplaceRequest, *xhttp.ParseError) {
```

Comment:
nit: maybe have this return a generic `error` since you're unwrapping it in the caller anyway? Fine either way

Reply:
Hmm, I see what you mean but since we use ParseError in all the other handler handlers I'm going to stick to local convention